chain/index/api.go

```diff
@@ -0,0 +1,345 @@
+package index
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+var ErrChainForked = xerrors.New("chain forked")
+
+func (si *SqliteIndexer) ChainValidateIndex(ctx context.Context, epoch abi.ChainEpoch, backfill bool) (*types.IndexValidation, error) {
+	// return an error if the indexer is not started
+	if !si.started {
+		return nil, errors.New("ChainValidateIndex called before indexer start")
+	}
+
+	// return an error if the indexer is closed
+	if si.isClosed() {
+		return nil, errors.New("ChainValidateIndex called on closed indexer")
+	}
+
+	// this API only works for epoch < head because of deferred execution in Filecoin
+	head := si.cs.GetHeaviestTipSet()
+	if epoch >= head.Height() {
+		return nil, xerrors.Errorf("cannot validate index at epoch %d, can only validate at an epoch less than chain head epoch %d", epoch, head.Height())
+	}
+
+	// fetch the tipset at the given epoch on the canonical chain
+	expectedTs, err := si.cs.GetTipsetByHeight(ctx, epoch, head, true)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset at height %d: %w", epoch, err)
+	}
+
+	// we need to take a write lock here so that back-filling does not race with real time chain indexing
+	si.writerLk.Lock()
+	defer si.writerLk.Unlock()
```

Comment:
```suggestion
	// we need to take a write lock here so that back-filling does not race with real-time chain indexing
	if backfill {
		si.writerLk.Lock()
		defer si.writerLk.Unlock()
	}
```

Reply:
any reason to hold a lock if we're just looking?

Reply:
This is done -> please feel free to resolve.